bin/generate-gitlab-ci-yml.sh

```diff
@@ -0,0 +1,80 @@
+#! /usr/bin/env bash
+
+original_directory=$(pwd)
+. "${original_directory}/share/spack/setup-env.sh"
+
+if [ -z "${DOWNSTREAM_CI_REPO}" ] ; then
+    echo "ERROR: missing variable: DOWNSTREAM_CI_REPO" >&2
+    exit 1
+fi
+
+if [ -z "${SPACK_RELEASE_ENVIRONMENT_PATH}" ] ; then
+    echo "ERROR: missing variable: SPACK_RELEASE_ENVIRONMENT_PATH" >&2
+    exit 1
+fi
+
+if [ -z "${CDASH_AUTH_TOKEN}" ] ; then
+    echo "ERROR: missing variable: CDASH_AUTH_TOKEN" >&2
+    exit 1
+fi
+
+if [ -z "${SPACK_RELEASE_ENVIRONMENT_REPO}" ] ; then
+    echo "Assuming spack repo contains environment" >&2
+    env_repo_dir=${original_directory}
+else
+    echo "Cloning environment repo: ${SPACK_RELEASE_ENVIRONMENT_REPO}" >&2
+    mkdir /work
+    cd /work
+    git clone ${SPACK_RELEASE_ENVIRONMENT_REPO} envrepo
+    cd envrepo
+    env_repo_dir=$(pwd)
+fi
+
+current_branch="$CI_COMMIT_REF_NAME"
+
+# Because want to see generated gitlab-ci file as an artifact,
+# we need to write it within the spack repo cloned by gitlab-ci.
+gen_ci_dir="${original_directory}/ci-generation"
+gen_ci_file="${gen_ci_dir}/.gitlab-ci.yml"
+mkdir -p ${gen_ci_dir}
+
+env_dir="${env_repo_dir}/${SPACK_RELEASE_ENVIRONMENT_PATH}"
+
+if [ ! -f "${env_dir}/spack.yaml" ] ; then
+    echo "ERROR: Cannot find spack environment file in ${env_dir}"
+    exit 1
+fi
+
+cd $env_dir
+
+token_file=/tmp/cdash_auth_token
+echo ${CDASH_AUTH_TOKEN} > ${token_file}
+
+# This commands generates the .gitlab-ci.yml and creates buildgroup in cdash
+spack release-jobs --force --output-file ${gen_ci_file} --cdash-credentials ${token_file}
+
+cp ${gen_ci_file} "${original_directory}/.gitlab-ci.yml"
+
+git config --global user.email "robot@spack.io"
+git config --global user.name "Build Robot"
+
+commit_msg="Auto-generated commit testing ${current_branch} (${CI_COMMIT_SHA})"
+
+cd ${original_directory}
+echo "git status"
+git status
+echo "git branch"
+git branch -D ___multi_ci___ 2> /dev/null || true
+echo "git checkout"
+git checkout -b ___multi_ci___
+echo "git add"
+git add .gitlab-ci.yml
+echo "git commit"
+git commit -m "$commit_msg"
+echo "git commit-tree/reset"
+git reset "$( git commit-tree HEAD^{tree} -m ${commit_msg} )"
```

Comment:
Might want to add a comment explaining what's going on with this line.

Reply:
At one point @tgamblin also suggested just pushing the single new commit as well.  Perhaps it's worth having a slightly more involved discussion about how we want to handle this? 

 Otherwise, yes, a comment will be nice here :smile: will do.

Reply:
It really boils down to whether the downstream repo is prepopulated with whatever is the main branch upstream.  In this case, sending history should be (much) faster than sending the working copy, since most of the WC is already there.  However, to maintain this benefit long term requires that the main downstream branch be continuously synchronized with the main upstream branch.  Otherwise (given enough history), sending history becomes much slower.

So, the real trade off made here is that we are always sending what amounts to a tarball of the WC, but in exchange we avoid the need for any external processes to keep the main branches synchronized across the two repos.  If there's an easy way to setup this synchronization (that doesn't require a paid GL license), then I'd be all for pushing the commits directly.

Maybe we could add that among the final "housekeeping" tasks we generate in CI?